lib/src/multires_stochastic_texture_synthesis.rs

```diff
@@ -1191,6 +1201,213 @@ impl PrerenderedU8Function {
     }
 }
 
+struct TreeGrid {
+    grid_width: u32,
+    grid_height: u32,
+    offset_x: i32,
+    offset_y: i32,
+    chunk_size: u32,
+    rtrees: Vec<RwLock<RTree<[i32; 2]>>>,
+}
+
+// This is a grid of rtrees
+// The idea is that most pixels after the first couple steps will have their neighbors close by
+impl TreeGrid {
+    pub fn new(width: u32, height: u32, chunk_size: u32, offset_x: u32, offset_y: u32) -> TreeGrid {
+        let mut rtrees: Vec<RwLock<RTree<[i32; 2]>>> = Vec::new();
+        let grid_width = width / chunk_size + 1;
+        let grid_height = height / chunk_size + 1;
+        rtrees.resize_with((grid_width * grid_height) as usize, || {
+            RwLock::new(RTree::new())
+        });
+        TreeGrid {
+            rtrees,
+            grid_width,
+            grid_height,
+            offset_x: offset_x as i32,
+            offset_y: offset_y as i32,
+            chunk_size,
+        }
+    }
+
+    #[inline]
+    fn get_space(&self, x: u32, y: u32) -> usize {
+        (x * self.grid_height + y) as usize
+    }
+
+    pub fn force_insert(&self, x: i32, y: i32) {
```

Comment:
Why "force" insert? I don't see a non-forceful version, and it's not clear what's forceful about this one.

Reply:
This is an artifact from before I removed the update queue when not every write inserted immediately. I will now just name it insert.